### PR TITLE
[polskieradio] adapt to next.js redesigns

### PR DESCRIPTION
**Pull request by selfisekai**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

adds new extractors that handle the redesigned pages (with new url schemes)

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

File: yt_dlp/extractor/polskieradio.py [hunk 1]
```diff
@@ -48,28 +50,11 @@ def _extract_webpage_player_entries(self, webpage, playlist_id, base_data):
...
-                'upload_date': '20160227',
-                'duration': 2364,
-                'thumbnail': r're:^https?://static\.prsa\.pl/images/.*\.jpg$'
-            },
-        }],
-    }, {  # New-style single broadcast.
+class PolskieRadioLegacyIE(PolskieRadioBaseExtractor):
+    # legacy sites
+    IE_NAME = 'polskieradio:legacy'
+    _VALID_URL = r'https?://(?:www\.)?polskieradio(?:24)?\.pl/\d+/\d+/[Aa]rtykul/(?P<id>[0-9]+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/polskieradio.py:56]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?polskieradio(?:24)?\.pl/\d+/\d+/[Aa]rtykul/(?P<id>\d+)'
```

File: yt_dlp/extractor/polskieradio.py [hunk 2]
```diff
@@ -48,28 +50,11 @@ def _extract_webpage_player_entries(self, webpage, playlist_id, base_data):
...
-                'id': '1540576',
-                'ext': 'mp3',
-                'title': 'md5:d4623290d4ac983bf924061c75c23a0d',
-                'timestamp': 1456594200,
-                'upload_date': '20160227',
-                'duration': 2364,
-                'thumbnail': r're:^https?://static\.prsa\.pl/images/.*\.jpg$'
-            },
-        }],
-    }, {  # New-style single broadcast.
```
Comment (pukkandan) [inline yt_dlp/extractor/polskieradio.py:72]: Why are the tests removed?

File: yt_dlp/extractor/polskieradio.py [hunk 3]
```diff
@@ -114,7 +89,10 @@ class PolskieRadioIE(PolskieRadioBaseExtractor):
     def _real_extract(self, url):
         playlist_id = self._match_id(url)
 
-        webpage = self._download_webpage(url, playlist_id)
+        webpage, urlh = self._download_webpage_handle(url, playlist_id)
+        # follow redirect if old url was supplied
+        if PolskieRadioIE.suitable(urlh.url):
+            return self.url_result(urlh.url, ie=PolskieRadioIE.ie_key(), video_id=playlist_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/polskieradio.py:95]:
```suggestion
        if PolskieRadioIE.suitable(urlh.url):
            return self.url_result(urlh.url, PolskieRadioIE, playlist_id)
```

File: yt_dlp/extractor/polskieradio.py [hunk 4]
```diff
@@ -153,23 +131,168 @@ def _real_extract(self, url):
         return self.playlist_result(entries, playlist_id, title, description)
 
 
-class PolskieRadioCategoryIE(InfoExtractor):
-    _VALID_URL = r'https?://(?:www\.)?polskieradio\.pl/\d+(?:,[^/]+)?/(?P<id>\d+)'
+class PolskieRadioIE(InfoExtractor):
+    # new next.js sites, excluding radiokierowcow.pl
+    _VALID_URL = r'https?://(?:[^/]+\.)?polskieradio(?:24)?\.pl/artykul/(?P<id>[0-9]+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/polskieradio.py:136]:
```suggestion
    _VALID_URL = r'https?://(?:[^/]+\.)?polskieradio(?:24)?\.pl/artykul/(?P<id>\d+)'
```

File: yt_dlp/extractor/polskieradio.py [hunk 5]
```diff
@@ -222,7 +342,10 @@ def _entries(self, url, page, category_id):
 
     def _real_extract(self, url):
         category_id = self._match_id(url)
-        webpage = self._download_webpage(url, category_id)
+        webpage, urlh = self._download_webpage_handle(url, category_id)
+        # follow redirect if old url was supplied
+        if PolskieRadioAuditionIE.suitable(urlh.url):
+            return self.url_result(urlh.url, ie=PolskieRadioAuditionIE.ie_key(), video_id=category_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/polskieradio.py:348]:
```suggestion
        webpage, urlh = self._download_webpage_handle(url, category_id)
        if PolskieRadioAuditionIE.suitable(urlh.url):
            return self.url_result(urlh.url, PolskieRadioAuditionIE, category_id)
```

File: yt_dlp/extractor/polskieradio.py [hunk 6]
```diff
@@ -153,23 +131,168 @@ def _real_extract(self, url):
...
+                i += 1
+
+    def _real_extract(self, url):
+        playlist_id = self._match_id(url)
+
+        page_props = traverse_obj(
+            self._search_nextjs_data(self._download_webpage(url, playlist_id), playlist_id),
+            ('props', 'pageProps'))
+        if page_props.get('data'):
+            page_props = page_props['data']
```
Comment (pukkandan) [inline yt_dlp/extractor/polskieradio.py:279]:
```suggestion
        page_props = traverse_obj(
            self._search_nextjs_data(self._download_webpage(url, playlist_id), playlist_id),
            ('props', 'pageProps', ('data', None)))
```

File: yt_dlp/extractor/polskieradio.py [hunk 7]
```diff
@@ -153,23 +131,168 @@ def _real_extract(self, url):
...
+                for episode in page['data']:
+                    yield {
+                        'id': str(episode['id']),
+                        'title': episode['title'],
+                        'url': episode['file'],
+                        'duration': int_or_none(episode.get('duration')),
+                        'timestamp': parse_iso8601(episode.get('datePublic')),
+                    }
+                not_empty = len(page['data']) > 0
+                i += 1
```
Comment (pukkandan) [inline yt_dlp/extractor/polskieradio.py:246]:
```suggestion
        for i in itertools.count(1) if has_episodes else []:
            page = self._call_lp3(
                'AudioArticle/GetListByCategoryId', {
                    'categoryId': playlist_id,
                    'PageSize': 10,
                    'skip': i,
                    'format': 400,
                }, playlist_id, f'Downloading episode list page {i}')
            if not traverse_obj(page, 'data'):
                break
            for episode in page['data']:
                yield {
                    'id': str(episode['id']),
                    'title': episode['title'],
                    'url': episode['file'],
                    'duration': int_or_none(episode.get('duration')),
                    'timestamp': parse_iso8601(episode.get('datePublic')),
                }
```

File: yt_dlp/extractor/polskieradio.py [hunk 8]
```diff
@@ -153,23 +131,168 @@ def _real_extract(self, url):
...
+                        '_type': 'url_transparent',
+                        'id': str(article['id']),
+                        'title': article['shortTitle'],
+                        'url': article['url'],
+                        'ie_key': PolskieRadioIE,
+                        'description': traverse_obj(article, ('description', 'lead')),
+                        'timestamp': parse_iso8601(article.get('datePublic')),
+                    }
+                not_empty = len(page['data']) > 0
+                i += 1
```
Comment (pukkandan) [inline yt_dlp/extractor/polskieradio.py:270]: same

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/polskieradio.py [hunk 9]
```diff
@@ -153,23 +130,166 @@ def _real_extract(self, url):
...
+            ('props', 'pageProps', ('data', None)))
+        details = page_props.get('details', {})
+
+        has_episodes = len(page_props.get('episodes', ())) > 0 or len(page_props.get('audios', ())) > 0
+        has_articles = len(page_props.get('articles', ())) > 0
+
+        return self.playlist_result(
+            self._entries(playlist_id, has_episodes, has_articles), playlist_id,
+            details.get('name'), traverse_obj(details, ('description', 'lead')),
+            thumbnail=details.get('photo'))
```
Comment (pukkandan) [inline yt_dlp/extractor/polskieradio.py:285]:
```suggestion
        has_episodes = bool(traverse_obj(page_props, 'episodes', 'audios'))
        has_articles = bool(traverse_obj(page_props, 'articles'))

        return self.playlist_result(
            self._entries(playlist_id, has_episodes, has_articles),
            playlist_id, title=traverse_obj(page_props, ('details', 'name')),
            description=traverse_obj(page_props, ('details', 'description', 'lead')),
            thumbnail=traverse_obj(page_props, ('details', 'photo')))
```

File: yt_dlp/extractor/polskieradio.py (hunk 2 above)
Comment (selfisekai) [reply yt_dlp/extractor/polskieradio.py:72]: moved from the legacy page extractor to the new frontend extractor below, as the site it was on got redesigned.
https://github.com/yt-dlp/yt-dlp/pull/5416/files/9190f6e541ec280ffa77b1dcde21a5462731275a#diff-7015cf2e8c40def45004b56bfcf2149d62f465abea11c89ddf817dab3c2f08b1R138

File: yt_dlp/extractor/polskieradio.py (hunk 6 above)
Comment (selfisekai) [reply yt_dlp/extractor/polskieradio.py]: this change makes `page_props` into an array with the object instead of the object, making the extraction fail

Comment (bashonly) [reply yt_dlp/extractor/polskieradio.py]: it needs `get_all=False`

File: yt_dlp/extractor/polskieradio.py [hunk 10]
```diff
@@ -153,23 +130,166 @@ def _real_extract(self, url):
...
+                        'skip': i,
+                        'format': 400,
+                    }, playlist_id, f'Downloading article list (page #{i})')
+                for article in page['data']:
+                    yield {
+                        '_type': 'url_transparent',
+                        'id': str(article['id']),
+                        'title': article['shortTitle'],
+                        'url': article['url'],
+                        'ie_key': PolskieRadioIE,
```
Comment (selfisekai) [inline yt_dlp/extractor/polskieradio.py:264]: hm, this doesn't seem to work? works with `.ie_key()`
```
ERROR: No suitable extractor (<class 'yt_dlp.extractor.polskieradio.PolskieRadioIE'>) found for URL https://jedynka.polskieradio.pl/artykul/1587425,Historia-zywa-nowy-cykl-w-radiowej-Jedynce
```

**Pull request merged**